composer/algorithms/agc/agc.py

```diff
@@ -0,0 +1,161 @@
+from composer.loggers import Logger
+
+log = logging.getLogger(__name__)
+
+__all__ = ["AGC", "apply_agc"]
+
+
+def apply_agc(
+    model: torch.nn.Module,
+    clipping_threshold: float = 0.01,
+    eps: float = 1e-3,
+) -> None:
+    """Clips all gradients in model based on ratio of gradient norms to parameter norms.
+
+    Example:
+         .. testcode::
+
+            from composer.algorithms.agc import apply_agc
+            apply_agc(model=model)
+
+
+    Args:
+        model (torch.nn.Module): The model being trained.
+        clipping_threshold (float, optional): The largest acceptable ratio between grad
+            norms and parameter norms before clipping is done.
+        eps (float, optional): Minimum value that weight norms are clamped to.
+    """
+    for param in model.parameters():
+        if param.grad is None:
+            continue
+
+        # Detach weights and gradients, so the clipping operation is not added to
+        # computational graph.
+        weights = param.detach()
+        grad = param.grad.detach()
+
+        # Get clipped version of gradients.
+        clipped_grad = _get_clipped_gradients(weights, grad)
+
+        # Copy clipped gradients into param.grad attribute, so they can be accessed by
+        # optimizer.
+        grad.copy_(clipped_grad)
+
+
+class AGC(Algorithm):
+    """Clips all gradients in model based on ratio of gradient norms to parameter norms.
+
+    From <https://arxiv.org/abs/2102.06171>.
+    Computes the norm of the weights and the norm of their corresponding gradients, then
+    scales the gradients by (weight_norm / grad_norm) * clipping_threshold for gradients
+    whose norms are greater than weight_norm * clipping_threshold. Norms are taken across
+    rows for weight matrices in MLPs, across entire filters/kernels for CNNs (channel and
+    spatial dimensions), and across the whole vector for biases.
+
+    Runs on ``Event.AFTER_BACKWARD``.
+
+    Example:
+         .. testcode::
+
+            from composer.algorithms import AGC
+            from composer.trainer import Trainer
+            agc_algorithm = AGC()
+            trainer = Trainer(
+                model=model,
+                train_dataloader=train_dataloader,
+                eval_dataloader=eval_dataloader,
+                max_duration="1ep",
+                algorithms=[agc_algorithm],
+                optimizers=[optimizer]
+            )
+
+    Args:
+        clipping_threshold (float, optional): The largest acceptable ratio between grad
+            norms and parameter norms before clipping is done.
+        eps (float, optional): Minimum value that weight norms are clamped to.
+    """
+
+    def __init__(self, clipping_threshold: float = 0.01, eps: float = 1e-3):
+        self.clipping_threshold = clipping_threshold
+        self.eps = eps
+
+    def match(self, event: Event, state: State) -> bool:
+        """Run on ``Event.AFTER_BACKWARD``."""
+        return event == Event.AFTER_BACKWARD
```

Comment:
In the trainer, we apply `torch.nn.utils.clip_grad_norm_` immediately before the AFTER_TRAIN_BATCH event. Is AFTER_BACKWARD the correct event for adaptive gradient clipping?

Reply:
Looking in the code, it looks like it gets called right before `AFTER_TRAIN_BATCH` and right after `self._train_microbatch` which runs `AFTER_BACKWARD`, so I think `AFTER_BACKWARD` might be the appropriate event?

Reply:
I think AGC will have to occur on `AFTER_TRAIN_BATCH`, basically in the same location as our current gradient clipping, because it has to be after microbatch grad accumulation, and after gradient unscaling.

Reply:
Oh good call, @abhi-mosaic! Also, sorry, @ravi-mosaicml, you said basically the same thing, but I misread it. Should be fixed now in 28dcddae23dc3f8ba5fccca3381acb32c3f3a692 . Thanks, folks!